File: problems/0007/0007-1.py
class Solution:
    def reverse(self, x: int) -> int:
        INT_MAX = (1 << 31) - 1
        INT_MIN = -INT_MAX - 1

        if x == INT_MIN or x == INT_MAX or x == 0:
            return 0
        signx = 1 if x > 0 else -1
        x = abs(x)
        res = 0
        while x:
            last_digit = x % 10
            res = 10 * res + last_digit
            if res > INT_MAX:
                return 0
            x //= 10
        return signx * res

File: problems/0001 - 0400/0007/0007-1.cpp
#include <climits>

class Solution {
public:
    int reverse(int x) {
        const int upper_bound = INT_MAX / 10;
        const int lower_bound = INT_MIN / 10;

        int res = 0;
        while (x) {
            int last_digit = x % 10;
            if (res > upper_bound || res < lower_bound) {
                return 0;
            }
            res = 10u * res + last_digit;
            x /= 10;
        }
        return res;
    }
};
